src/configuration.ts

```diff
@@ -36,86 +45,319 @@ export class AvmDebugConfigProvider implements vscode.DebugConfigurationProvider
+    const programSourcesDescriptionFile = await this.getProgramSourcesDescriptionFile(folder, config)
+    const simulateTrace = await getSimulateTrace(config.simulateTraceFile)
+    if (!simulateTrace) return null
 
-    if (!programSourcesDescriptionFile) {
-      const workspaceFolder = folder
-      const sourcesUris = await findFilesInWorkspace(workspaceFolder, SOURCES_FILE_PATTERN)
+    const uniqueHashes = getUniqueHashes(simulateTrace)
+    const sources = await this.getSources(programSourcesDescriptionFile)
+    const missingHashes = getMissingHashes(uniqueHashes, sources)
 
-      if (sourcesUris.length === 1) {
-        programSourcesDescriptionFile = sourcesUris[0].fsPath
-      } else {
-        const pickedSource = await vscode.commands.executeCommand<string | undefined>(
-          'extension.avmDebugger.getProgramSourcesDescriptionFile',
-          config,
-        )
-        if (!pickedSource) {
-          return null
+    if (missingHashes.length > 0) {
+      await this.handleMissingHashes(folder, missingHashes, simulateTrace, sources)
+      await this.persistSources(folder, programSourcesDescriptionFile, sources)
+    }
+
+    const curatedProgramSourcesPath = await this.createCuratedProgramSources(sources, uniqueHashes, programSourcesDescriptionFile)
+    return { ...config, programSourcesDescriptionFile: curatedProgramSourcesPath.fsPath }
+  }
+
+  private async createCuratedProgramSources(
+    sources: SourcesFile,
+    uniqueHashes: string[],
+    programSourcesDescriptionFile: string,
+  ): Promise<vscode.Uri> {
+    const PRIORITY_MAP: Record<string, number> = {
+      '.puya.map': 3,
+      '.teal.map': 2,
+      '.tok.map': 1,
+      '': 0,
+    }
+
+    const getPriority = (source: SourceRecord): number => PRIORITY_MAP[path.extname(source['sourcemap-location'] ?? '')] ?? 0
+
+    const curatedProgramSources = (sources['txn-group-sources'] ?? [])
+      .filter((source): source is SourceRecord => source !== undefined && uniqueHashes.includes(source.hash))
+      .reduce<SourceRecord[]>((acc, source) => {
+        const existingIndex = acc.findIndex((s) => s.hash === source.hash)
+        if (existingIndex === -1) {
+          acc.push(source)
+        } else if (getPriority(source) > getPriority(acc[existingIndex])) {
+          acc[existingIndex] = source
+        }
+        return acc
+      }, [])
+
+    const curatedProgramSourcesPath = vscode.Uri.joinPath(
+      vscode.Uri.file(programSourcesDescriptionFile),
+      '..',
+      `${Date.now()}.sources.avm.json`,
+    )
+
+    // Write curated sources to the temporary file
+    await vscode.workspace.fs.writeFile(
+      curatedProgramSourcesPath,
+      Buffer.from(JSON.stringify({ 'txn-group-sources': curatedProgramSources }, null, 2)),
+    )
+
+    // Register cleanup function
+    this.disposables.push(
+      vscode.debug.onDidTerminateDebugSession(async () => {
```

Comment:
I'm curious if this also runs when the user kills vscode when actively debugging or in other potentially non graceful scenarios.

Reply:
no longer the case as we removed the approach that relies on tmp files for curated sources